Adds dimension controls to Group Block

getdave opened the pull request:
Closes https://github.com/WordPress/gutenberg/issues/14747

![Screen Capture on 2020-02-24 at 15-43-05](https://user-images.githubusercontent.com/444434/75166871-76ad6f80-571c-11ea-998c-fa2333672753.gif)


## Description
Adds dimension controls (currently limited to `padding`) to the Group Block via experimental `DimensionControl` component within the `@wordpress/components` package.


**Please note**: this PR aims to implement the basic feature set for dimension controls. More complex UI will be added at a later date. For a more detailed explanation see [this comment](https://github.com/WordPress/gutenberg/pull/16730#issuecomment-515211211).

## Todo

- [x] Ensure i18n best practices are observed
- [x] Unit tests
- [ ] Add deprecation to `core/group` Block
- [x] Add `reset` UI
- [ ] ~Add individual padding side control~ (probably out of scope)
- [ ] ~Enable responsive settings~ (descoped as superceeded by other work in other PRs)

## Screenshots




## How has this been tested?
Manually for now.

To test please use [these test annotations](https://gist.github.com/getdave/a30705607716d08c47046a45a6758c1d).

1. Create a new Page
2. Switch to `Code Editor` mode from the editor settings menu (dots in topmost right corner).
3. Paste in all [the test annotations](https://gist.github.com/getdave/a30705607716d08c47046a45a6758c1d).
4. Switch back to `Visual editor` (again using editor settings menu top right).
5. See all possible variations of dimension combinations.
6. Add backgrounds to make things clearer.

## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

getdave committed:
Adds basic dimension controls to Group Block

getdave committed:
Moves Dim Control to block editor package. Makes size class based rather than hard coded value

getdave committed:
Moves sizes config to seperate file. Introduces tooltip on size buttons

getdave committed:
Adds aria pressed support to size buttons

Soean commented on packages/block-editor/src/components/dimension-control/index.js:
```diff
@@ -0,0 +1,110 @@
+	return (
+		<BaseControl
+			id={ controlId }
+			help={ `Select the ${ type } for this Block` }
```
This Sting is not translatable.

getdave replied:
Thanks. Good catch.

> The type should also be translatable.

I disagree here. This represents a formal CSS property value and thus shouldn't be available for translation. Perhaps we should allow a custom value to describe the property in addition to the formal `prop`?

Pushed some fixes.

getdave committed:
Improve i18n of strings.

Addresses https://github.com/WordPress/gutenberg/pull/16730#discussion_r306758167

Renames `type` to `property` to better indicate that it is a formal CSS property value and therefore should not be translated lest it break the underlying code.

Added title attribute to enable true text-based strings to be translatable.

paulwilde commented:
Maybe `Extra Large` would make more sense instead of `Huge`. I didn't know what "H" stood for at first.

getdave committed:
Renames Huge to Xlarge and adds ability to provide abbr for a11y reasons

getdave committed:
Adds reset button to UI

getdave committed:
Align UI closer to mockup provided by @kjell

See https://github.com/WordPress/gutenberg/issues/13363#issuecomment-463291100

getdave commented:
@kjellr has pointed me at some nice UI for responsive controls. I'll be implementing that

https://github.com/WordPress/gutenberg/issues/13363#issuecomment-463291100

getdave committed:
Add Responsive Controls and Attributes

getdave committed:
Move Group Inspector to dedicated component

getdave committed:
Remove dependence on clientID to make component more portable and generic.

getdave committed:
Adds doc blocks

getdave committed:
Moves findSizeBySlug util function into sizes file

getdave committed:
Adds basics tests

getdave committed:
Adds component docs

getdave committed:
Update to decouple DimensionControl component from knowledge of Block attribute schema

Previously the `DimensionControl` component manually constructed the attribute to be updated via string concatenation. This was brittle and prone to error and also couple the component to use in a Block context.

Updated to utilise partial application to apply the dimension attribute argument in advance for each component. This allows us to simply the API of the component.

paulwilde commented on packages/block-library/src/group/inspector.js:
```diff
@@ -0,0 +1,232 @@
+						<DimensionControl
+							title={ __( 'Padding' ) }
+							property="padding"
+							id={ clientId }
```
You can ease the requirement to constantly pass a unique `id` into the component by using the `withInstanceId` HOC which other controls are already using:

https://github.com/WordPress/gutenberg/blob/9fcccb3c374238a161fc92dc414f0e11d135606e/packages/components/src/text-control/index.js#L29

getdave replied:
Great advice. I'd forgotten about that one.

Annoyingly adding this has broken all my tests because even with Enzyme `mount` doesn't render the `<button>` elements now they are wrapped in a HOC.

getdave replied:
Fixed.

SchneiderSam commented:
Nice, but I find the information with S, L, M, XL ect somewhat strange. Why not with pixels? The current variant is nice for beginners, but very inaccurate for advanced users...

drw158 commented:
Based on https://github.com/WordPress/gutenberg/issues/13363#issuecomment-476714596

Some design feedback:

- The switch help text shouldn't be phrased as a question. Also, I don't think margin and padding should be capitalized in the switch help text.
- The help text under each of the controls is redundant (`Select the padding for this block`).
- The labels could be simplified. Try `Mobile` instead of `Padding (mobile devices)`.
- I assume the double arrow icon on the right resets the setting but I'm not sure. Might be better to use a text button that reads "Reset".
- I don't know what `N` and `X` do.

getdave commented:
> I find the information with S, L, M, XL ect somewhat strange.

@SchneiderSam Thanks for your opinion. I assume by "strange" you mean that it wasn't immediately clear to you why the arbitrary values had been selected over absolute units? 

> Why not with pixels?

Imagine you are a Theme designer. You craft your CSS with spacing that is perfect for the design. You want to ensure that is consistent throughout your Theme, even if the page layout is being created by the end-user in the Block Editor.

With the approach I've taken here, when a size is selected only classes are added to the Block in the DOM. This affords the Theme creator the opportunity to provide custom sizes in CSS that are suitable for their Theme. If they opt not to do this then sensible defaults are provided.

With the pixels approach, we're locking users of the Block into absolute values and asking them to make a lot of decisions that they'd probably prefer not to have to make. It could also lead to an inconsistent visual experience. 

> The current variant is nice for beginners...

I completely agree. This is why I've selected this approach for the first pass at adding this control. Average users (whoever they are!) will likely be grateful to receive a series of _predefined_ options rather than having to make their own choices. I'm looking to cater to them first, which also serves to keep this PR simple and easily comprehensible, this increasing the changes it will be accepted and merged.

> ...but very inaccurate for advanced users...

I take your point. As a power-user, you want _precise control_ which this approach doesn't _currently_ provide.

Please consider however, that at this point we're introducing a new feature with a totally new UI. As a result, I'm keen to avoid overcomplicating this PR or increasing its scope.

If this UI ships, in the future, I imagine we will want to enhance it with an "advanced" option which will afford power users the ability to 

- specify absolute values
- specify the values for each "side" _individually_ (eg: `top`, `left`...etc)

Ultimately, there's a lot of things we _could_ look to do. However, for now, I'm hoping you will agree that keeping things simple and focus on the 90% of users will afford us the best opportunity of getting this PR merged. We can iterate from there.

Once again, I want to say how much I really appreciate you taking the time to provide feedback.

getdave commented:
@drw158 Thanks for your feedback! All good stuff. I'll look at updating it tomorrow.

> I don't know what N and X do.

Note that there are tooltips for each option with the full text available.

I think `X` has now become `XL` which should solve things. Do you agree that is clear enough?

I agree that `N` is a problem. I'd value any alternatives. Bear in mind an explicit `None` option is required because some Blocks have default padding and so we need `None` as an option to allow users to completely remove the padding from the Block.

paaljoachim commented:
> I agree that N is a problem. I'd value any suggests for alternatives. Bear in mind an explicit None option is required because some Blocks have default padding and so we need None as an option to allow users to completely remove the padding from the Block.

What about a No with a tooltip of None.

getdave committed:
Utilise withInstanceId HOC to remove need for id prop

Also updates tests to account for this.

getdave committed:
Updates size buttons for improved comprehension and a11y

Updates `N` to be `No`. Also avoids string manipulation approach to getting visual size button label in favour of explicit abbr prop on sizes.

getdave commented:
> What about a No with a tooltip of None.

> I don't know what N and X do.

@drw158 @paaljoachim I've updated to the below (`No` and `XL`) and also improved the a11y

<img width="268" alt="Screen Shot 2019-07-26 at 10 32 26" src="https://user-images.githubusercontent.com/444434/61942269-fa3d8d00-af90-11e9-81c9-e0480c6f09bf.png">

Is this better?

paaljoachim commented:
I shared the PR in the #design channel on Slack to get some additional views.
@mapk

getdave commented:
@drw158 I've fixed the below based on your feedback:

> The switch help text shouldn't be phrased as a question. Also, I don't think margin and padding should be capitalized in the switch help text.

✅ Updated.

> The help text under each of the controls is redundant (Select the padding for this block).

✅ Agreed. This was a hangover and is redundant. Removed.

> The labels could be simplified. Try Mobile instead of Padding (mobile devices).

✅ I think this removes a lot of noise from the UI. Updated.

> I assume the double arrow icon on the right resets the setting but I'm not sure. Might be better to use a text button that reads "Reset".

✅ Yep that's right. I've converted to a true `Button` with text and also disabled when there is no value currently selected.

> I don't know what N and X do.

✅ Addressed [above](https://github.com/WordPress/gutenberg/pull/16730#issuecomment-515380997).

Here's a screenshot for reference:

<img width="274" alt="Screen Shot 2019-07-26 at 10 56 19" src="https://user-images.githubusercontent.com/444434/61943765-11ca4500-af94-11e9-99a7-0c4596705563.png">

getdave committed:
Updates UI in response to designer feedback

See https://github.com/WordPress/gutenberg/pull/16730#issuecomment-515389101

bharath commented:
Hi, I have some thoughts to share on this.

I agree predefined options are a good start and specifying absolute values as stated can be added as an advanced option later.

But I think the option to specify those predefined options for each "side" individually should also be there from the start. 

A lot of themes have full-width sections where only padding on the top and bottom is needed. 

And also sometimes different values are used on the top and bottom.

mtias commented:
Thanks for working on this @getdave. We already experimented with a similar UI for font sizes and switched away to a dropdown menu due to lack of clarity and internationalization issues. Can we try that here?

mtias commented:
Also, I'd be inclined to separate this into two PRs because the device specific sizes need a bit more thinking across the entire UI.

karmatosed commented:
> Also, I'd be inclined to separate this into two PRs because the device specific sizes need a bit more thinking across the entire UI.

I would plus one this as it's getting huge. I also agree with the selection of sizes. Overall this feels a lot to take in visually and I think there should be a step back to consider what this would be like for someone coming to it. Are they going to understand all these settings? Are we using the right terms?

A prime example of how much this becomes is when you show all devices and this is what see on my 13 inch screen (I would have to scroll):

![image](https://user-images.githubusercontent.com/253067/61951025-94a8cb00-afa7-11e9-8cea-7e0be5eb23af.png)

I was very confused by the following icon and it didn't seem to do anything.

![Artboard2](https://user-images.githubusercontent.com/253067/61951045-a12d2380-afa7-11e9-9344-17c894f56844.png)

Overall there is some key spacing to adjust and this demonstrates visually how this is all a lot:

![Artboard](https://user-images.githubusercontent.com/253067/61951257-62e43400-afa8-11e9-9e55-44f7c5b784ce.png)

kjellr commented on packages/block-library/src/group/theme.scss:
```diff
@@ -6,4 +6,181 @@
 		margin-top: 0;
 		margin-bottom: 0;
 	}
+
```
This is going to be really difficult for themes to override if they need to: there are a ton of different breakpoint/margin/padding combinations, and I worry that the theme would essentially need to rewrite all of them. This _should_ work out of the box in many themes (it seems to be okay in Twenty Nineteen based on my testing), but will definitely break some too. For instance, allowing users to adjust the margins in the Gutenberg starter theme causes it to completely fall apart, since it relies on all blocks having `auto` left and right margins. For that reason, I wonder if this is this "Allow users to control margin and padding for individual blocks" is something that themes opt-in to? Similar to how they opt into wide alignments currently. 

On a related note, most themes won't actually use the exact same breakpoints we use here. In #13363, we talked about potentially allowing themes to specify breakpoints for each of these devices. I'm not sure how/if that would best be implemented though. 🤔

getdave replied:
> This is going to be really difficult for themes to override if they need to

I've got to be honest, I've kind of hacked this in, in the hope that someone more famliar with Themes can advise on how best to implement to keep specificity down and allow Themes to extend. I'd greatly value any commits you'd like to push in this direction.

> On a related note, most themes won't actually use the exact same breakpoints we use here

Oh god, I didn't think of that. Of course, the Block Editor uses standard breakpoints, but Themes will choose their own.

I agree that we need a way to allow users to opt-in and to specify breakpoints. This seems like a wider project. 

What route do we have to allow an initial MVP implementation of Block spacing to be shipped? Or do we need to get the breakpoint registration stuff done first?

Thanks!

kjellr replied:
> I've got to be honest, I've kind of hacked this in, in the hope that someone more famliar with Themes can advise on how best to implement to keep specificity down and allow Themes to extend. I'd greatly value any commits you'd like to push in this direction.

I'll have to think about this a little bit. I wonder if @allancole, @laurelfulford, or @davidakennedy has any thoughts too. In general, I think we'll have much more leeway if we make this an opt-in for themes. 

> What route do we have to allow an initial MVP implementation of Block spacing to be shipped? Or do we need to get the breakpoint registration stuff done first?

I wonder if shipping the automatically-responsive spacing makes the most sense as an MVP. Then we can tackle the breakpoint-specific settings alongside theme-supplied breakpoints in a separate, wider issue.

getdave replied:
> I'll have to think about this a little bit

Thanks 👍 

> I wonder if shipping the automatically-responsive spacing makes the most sense as an MVP. Then we can tackle the breakpoint-specific settings alongside theme-supplied breakpoints in a separate, wider issue.

Agreed and I think this is what I will do. So MVP will ship with basic controls (no responsive). Then I'll work on branches for 

* a generic <ResponsiveControls /> component (as seen here but not extracted)
* the `<DimensionControl />` component (as seen here)

Then when we've had a chance to think about the impact of responsive settings on Themes we can look to bring it all together.

ZebulanStanphill replied:
Why is this in the `theme.scss` file if it is an editor-only style?

getdave commented:
> Thanks for working on this @getdave.

@mtias Thanks for this. Much appreciated.

> We already experimented with a similar UI for font sizes and switched away to a dropdown menu due to lack of clarity and internationalization issues. Can we try that here?

Yep, that makes sense. I'm starting to think this might be a good way to reduce the visual noise of the UI.

The main downside I can see is that it would require x2 clicks to toggle between different sizes. It is nice to be able to just click around and see the Block update in real time. The dropdown will mean a lot of clicks to experiment with different sizes. 

On balance however, I suspect we'll end up going with the dropdown for the reasons you've stated.

> Also, I'd be inclined to separate this into two PRs because the device specific sizes need a bit more thinking across the entire UI.

@mtias Can I confirm you're recommending

* PR for a UI for Responsive controls (toggle to show/hide responsive) 
* PR for Dimension controls
* PR for adding spacing controls to the Group Block

==========

> Overall this feels a lot to take in visually...

@karmatosed I agree. As a non-designer, I'd definitely value and appreciate guidance from more visually talented folks.

> ...and I think there should be a step back to consider what this would be like for someone coming to it.

I agree and this is the moment to do that. I've demonstrated one implementation based on @kjellr's visuals. I'd love to receive alternative suggestions and options. 

> A prime example of how much this becomes is when you show all devices and this is what see on my 13 inch screen (I would have to scroll):

It's certainly something to consider. I agree there is a lot to take in. However, having everything open is something of an advanced use case. Indeed, I'm not sure it is that terrible to see a lot of controls, especially as they are clearly labelled and divided up. 

I also think it's preferable for controls to visible by default rather than (unnecessarily) hiding the responsive UI controls behind toggles/switches as some other Block frameworks do. I'd rather the "noise" than make the UI overly complex or unintuitive.

We could reduce the noise by adopting the dropdown/select approach that @mtias suggests. Please see my comments on that above.

Other than that, do you have any alternatives we could try to reduce the visual noise? Much appreciated.

> I was very confused by the following icon and it didn't seem to do anything.

I think you're looking at an out of date branch. Pull down the latest changes to see the updates. 

You will see that - as per @drw158's comments - the reset UI is now clearly labelled and only active once a size is selected. I agree it was previously confusing 😄 


Also, happy to fine-tune the spacing and alignment once the core UI has been agreed. Based on previous experience, I'm sure you'll understand that I'm disinclined to commit time to fine-tuning UI that hasn't got general approval from the community. I agree this is important to get right however 👍 

Thanks for taking the time to feedback. It's much appreciated.

karmatosed commented:
> However, having everything open is something of an advanced use case. Indeed, I'm not sure it is that terrible to see a lot of controls, especially as they are clearly labelled and divided up.

Whilst yes to interact would be advanced it was pretty easy and pretty much an expected interaction pattern got me to them all being open. In this case I am not sure it is clear how things are labelled and divided up, so that is absolutely something to work on.

getdave commented:
> But I think the option to specify those predefined options for each "side" individually should also be there from the start.

@bharath Thanks for this. I understand why you bring it up. It's difficult to balance ideal functioning vs essential functionality. I feel we should be able to introduce this fairly easily in future PRs once the initial implementation is complete. Again, it's about not over-complicating the initial implementation (I say this as the person who has a PR with 19 files changed! 😄 ).

@mtias What say you to this?

getdave commented:
> Whilst yes to interact would be advanced it was pretty easy and pretty much an expected interaction pattern got me to them all being open. In this case I am not sure it is clear how things are labelled and divided up, so that is absolutely something to work on.

@mtias @karmatosed Whilst still in this branch, I've been experimenting with using select controls as per Matias' suggestion. Here's a screenshot of this WIP. It does remove a lot of visual noise.

<img width="288" alt="Screen Shot 2019-07-26 at 15 33 40" src="https://user-images.githubusercontent.com/444434/61959257-edce2a00-afba-11e9-92fa-cf020d176b4a.png">

Do you feel this is an improvement?

getdave committed:
Update to utilise dropdown for UI. Improve alignment. Move toggle below inputs.

getdave commented:
Following multiple requests, this is being broken up into separate PRs.

See [description](https://github.com/WordPress/gutenberg/pull/16730#issue-300659967).

paaljoachim commented:
Hey Dave @getdave 
it would be great with an update to where this is at and how it can be moved forward.
Thanks!

getdave commented:
Hey. I've been away. Planning to try and give this some attention soon.

getdave commented:
Update: still working on sub PRs (see description). Once done these will be merged into this PR and work will recommence here.

getdave commented:
Update: the two sub-PRs are nearly merged. Once done, this PR will be rebased and updated to utilise the new components. We can then look at wiring this together again.

getdave committed:
Remove duplicate Dimension Control

This was handled in a separate PR and resides in `@wordpress/components`. Can be removed from this PR.

getdave committed:
Refactor dimension controls to use new experimental components

getdave committed:
Removes Dimension Control docs

This was automated.

getdave committed:
Updates to handle resetting values when toggling responsive mode.

getdave committed:
Adds icons to padding controls

getdave committed:
Remove “screens” from viewport labels as already in aria description

getdave committed:
Fix to ensure labels and fields align right hand edges correctly

getdave commented:
@mapk @kjellr @jorgefilipecosta @jasmussen  I've merged both sub PRs and rebased and refactored this component to utilise these. 

I think the main issue now is how we provide default styles both within the editor and on the Theme frontend without causing problems.

When a value is selected a class is added to the Group Block. But what sensible defaults can we provide? How will this work with developers who want to override these?

I'd appreciate any help/advice to move this towards completion.

obenland commented:
I'm not sure how workable this PR is meant to be at this point, so take this with a grain of salt. In my testing it didn't seem to add any size classes in either the editor and the front-end when adjusting screen-size-based paddings, so I couldn't observe any changes in padding. Or I just did it wrong :)

It felt counterintuitive having to disable a setting to unlock finer-granular settings. Is there something I'm missing?

hacknug commented:
Will we get a PHP API to define our own `viewports` for themes?

getdave commented:
> Will we get a PHP API to define our own viewports for themes?

@hacknug That's a very good question. It's not something I'm going to tackle in this PR. It would be great if you would consider [making an `Issue` for that](https://help.github.com/en/github/managing-your-work-on-github/creating-an-issue).

-----------

@obenland I feel like I may have broken something and not realised. I'm going to check in with this again soon and ensure it's working before recalling your attention to it.

> It felt counterintuitive having to disable a setting to unlock finer-granular settings. Is there something I'm missing?

This has been back and forth a few times already. We've settled on the current mechanic of the toggle. I'd prefer not to reopen that question if possible 😄.

jasmussen commented:
I was just working on a block that includes a way to set separate dimensions on a per-breakpoint basis. While we ended up going with an interface not unlike this, i.e. separate breakpoint controls in the sidebar, we've already tracked this to be improved in any way we can. The problems with it are:

1. There's no preview of what it does, at all. You can flip a switch and nothing happens. 
2. Because there's no live preview, the interface in the sidebar is needlessly complicated with several rows of controls. 

The second item has a compounding effect that kind of ruins the contextual nature of the sidebar: what if I want a dropcap on desktop but not on mobile? What if I want fullwide on mobile and tablet but not on desktop? This is probably the most problematic aspect: any control we want to add responsiveness to has to be duplicated endlessly inside these dimension controls.

As such, I deeply appreciate the work that's gone into this and how it's helped uncover these challenges. But as much as it pains me to say, maybe that's also where it should end. I worry that the same interface will be duplicated across blocks, sometimes on a per-feature basis, begging the question what goes in these controls and what doesn't. 

This does not mean the problem at hand shouldn't be addressed, but that maybe it should be addressed differently. I was very inspired to see [this comment](https://github.com/WordPress/gutenberg/issues/13203#issuecomment-550389126) by @aduth. The technical side is complex, but the principle is right: if you need to adjust block parameters on a per-viewport basis, you must be able to see those parameters have an instant effect on the block itself.

getdave commented:
@jasmussen Thanks for your feedback. This PR has been open for some time and has spawned some useful components. I take your points on the problems of this UI and personally I'm happy to drop this approach in favour of the ones you've referenced and are exploring. 

I appreciate your candid feedback.

getdave commented:
@jasmussen @mapk Do you still think that adding the padding/margin control to the Block _without_ responsive controls has value? I note [the comments on this PR](https://github.com/WordPress/gutenberg/issues/11824).

greylockglass commented:
Just an end-user here with zero coding skills, so understand that when I ask for a pony and a time machine, I don't actually know what goes into providing them. That said, after reading through this thread, most recently @jasmussen 's comment, I do understand the basic can-of-worms that is opened up by adding padding controls. Would it make any sense, or no sense at all, to have a separate block (The Adjustable Padded Cell Block) with these controls? That way, only users who have such a need would have to figure it out? Also, I am firmly in favor of keeping the Spacer block. I build some moderately complex post layouts, and, pre-Gutenberg, used an image of a fancy horizontal rule to force section breaks in my content (html for horizontal rule always seemed to get kicked out). Some time after I switched to Gutenberg, I needed a spacer and Googled "Is there a spacer block in Gutenberg?" and moments later, I was happily separating blocks of content again, more easily than ever. Just my 2¢. Thanks!

mapk commented:
Based on [this comment](https://github.com/WordPress/gutenberg/issues/11824#issuecomment-570179526) from @getdave: 

>However, if we still feel that adding padding/margin support to the Group block (without responsive support for now) is useful I'm still happy to pursue it.

I think this is still valuable. While I agree with [Joen's comment above](https://github.com/WordPress/gutenberg/pull/16730#issuecomment-551666362), maybe we can still add these controls for the block overall without responsive support.

I've personally run into this problem several times in regards to individual blocks and have [heard from others](https://github.com/WordPress/gutenberg/issues/11824) that they'd like more control over padding/margins on a per block basis.

jasmussen commented:
Thanks for all the positive energy here. I'm only just back from vacation so hence the slow response, but it looks like everything ended in a good place. 

As a followup, I have some mockups that I'm still tweaking a bit, that I hope to post very soon on a ticket. In those mockups, the responsive support is being added _separately_ from the individual option, and as a result of first choosing the viewport resizing. You can see an early version of those mockups in this comment: https://github.com/WordPress/gutenberg/pull/19082#issuecomment-567508590

In other words, yes, if padding is added to the group without responsive support, the responsive support should be possible to land separately without extra work.

getdave committed:
Removes responsive setup

getdave committed:
Remove all responsive CSS styles

getdave committed:
Add margin and i18n

getdave commented:
# Important Update

Following comments (see above) this PR has been descoped to handle margin/padding without attempting to handle the responsive settings for these. The responsive settings will be handled via more sophisticated mechanisms currently under development in other PRs.

getdave committed:
Adds help text to Dimension Control block

getdave committed:
Adds help notice to Margin controls.

getdave committed:
Update snapshots.

WunderBart commented:
> Is this better?

Out of curiosity, would eg. a `∅` (null symbol) instead of `No` be an option here? @getdave

getdave committed:
Updates to coolocate all attributes in block.json

getdave committed:
Update full content integration tests

getdave commented:
@mapk I think I need some input from Design here. Main points:

1. Does the Editor experience feel "right" in general?
2. Is it ok to only allow margin to control horizontal spacing (as previously suggested)?
3. Is [using multiples of the `$block-padding` variable](https://github.com/WordPress/gutenberg/pull/16730/files#diff-a1678e3d2dad77286726884a772e9940R10) the best approach in terms of styling?

I'd like to get this wrapped up so any support you (or others) can offer would be appreciated.

karmatosed commented:
I tested this a little and whilst it does work there is a bit of a confusing cognitive space here. I think we might be able to solve some of this with the copy but overall there's a confusion between what padding and margin mean. Similarly, not seeing what 'large' means is hard to know what it will do. For example, I didn't seem to be able to get padding working on a group containing an image and paragraph.

Text here is also a little confusing, what is horizontal margin and if we are recommending the vertical spacer, maybe we can advise in a way here that guides a bit more. I'll add a copy label to see if we can get some support for this.

My biggest issue with it is there seems a lot of guessing and unexpected outcomes. I don't know what a change is going to bring or why I would click. Again, I think this can have copy and maybe some adjusting in the way this is presented to clarify that. For example, calling it spacing but recommending you use a spacer block as it won't deal with that, maybe just call it 'Horizontal spacing' if that's all that can be done.

How possible is it to be a slider you can actually see real-time the changes, over a drop-down? There could be something to that like columns.

@getdave to answer specifically 2 and 3, let me loop in @ItsJonQ and @jasmussen along with you've already looped in Mark, let's get some eyes on this for you.

michelleweber commented:
@karmatosed Happy to offer whatever thoughts I can on copy. Can you post or link me to what screenshot I should be looking at? It's quite a long convo and I want to make sure I give you feedback on the right thing!

mapk commented:
I tested this out @getdave! The controls are very simple and feel pretty good. I agree with @karmatosed that we definitely need to revise some copy though to help convey what's going on.

Just for kicks, here's what it looks like and what's going on in the code.

![margin-padding 2020-01-07 15_32_34](https://user-images.githubusercontent.com/617986/71939487-ac47ca80-3167-11ea-84a5-04e9307de547.gif)

## Feedback

### 1. The vertical margin
While I understand why there's a note about vertical margin, it's still actually increasing the CSS margin both on top and bottom when I change the setting... even though it's not visible. So my recommendation is maybe we don't need that text. It's just an overlapping of margins that seems to be going on which is why we're not seeing very much change there, correct? I'm not sure what the previous comment was about this.

Instead of the current wording:

> Controls horizontal margin only. Please use the Spacer block for vertical spacing.

Maybe we write it differently?

> Vertical margins may vary due to CSS margin overlapping and should be reflected better on the frontend. 

Yea, that's not the best either... but more to the truth of the matter.

### 2. Frontend
I didn't notice any change in the frontend yet. Is that correct? If I'm right that the frontend work just hasn't been done yet, it might make a difference to number 1 above.

### 3. Immediate visual response 
I was able to see immediate visual changes with every selection – really good stuff!

> My biggest issue with it is there seems a lot of guessing and unexpected outcomes. I don't know what a change is going to bring or why I would click.

I think this ^ is okay, @karmatosed. It's kinda like selecting a font. I'm not sure what the page is going to look like until I actually select the font and it applies across the board. So waiting to see an actual change _after_ the dropdown selection is made feels right to me. It does require experimentation on behalf of the user, but so does most everything else like color selection, etc.

### 4. Margin & Padding
Are those words clear enough for people? Do people know the difference? Would we want to include any graphic to help explain this like:

![margin-padding](https://user-images.githubusercontent.com/617986/71940167-2e38f300-316a-11ea-8d60-d739f3a6a205.png)

This graphic really isn't accessible or translatable, so it's probably a super bad idea. But I wanted to throw it into the mix anyways. Don't let it hold up this PR.

getdave commented:
@michelleweber [The screencapture in @mapk's comment ](https://github.com/WordPress/gutenberg/pull/16730#issuecomment-571835512)now shows the entire UI. I'll try and get around to doing a static screengrab for you.

michelleweber commented:
Thoughts:
- I feel like we need to define margins/padding for people with some explanatory text, at a least a little. Maybe something like "Margins (adjust spacing on the sides of the block)" and "Padding (adjust spacing around content within the block)." I don't know if the text goes under the terms themselves or if both definitions go at the bottom of the settings; that's a question for y'all.
- For the "vertical" point, I'd probably try to keep it really simple in the main settings, like "To add space above or below a block, insert a Spacer block.

The question for me is, what language is in the users' heads? Are people thinking "I want to change my horizontal padding" or are people thinking "I wish there was less space on the sides of that block"?

More than that seems like too much for the basic settings, esp anything about vertical margins and CSS -- that's only going to be meaningful to a wee group of people, and intimidating to everyone else.

getdave commented:
Thanks for the excellent and timely feedback @mapk @karmatosed and @michelleweber. Very much appreciated. Will be working on this again and will ping when ready for re-review.

getdave committed:
Remove unwanted help text

Implements feedback from https://github.com/WordPress/gutenberg/pull/16730#issuecomment-571835512

getdave committed:
Restore enhancements lost during rebase

getdave committed:
Revert unwanted change to Responsive Controls

getdave committed:
Remove random new line

getdave committed:
Remove deprecated file reintroduced via rebase

getdave committed:
Adds help text to controls

Addresses https://github.com/WordPress/gutenberg/pull/16730#issuecomment-572089124

getdave committed:
Add padding and margin to saved Block for use on frontend

getdave committed:
Add subtle animation to margin/padding within Editor only

getdave commented:
Sorry for the delay here. Been sidetracked onto the Nav Block for the past month and a bit.

> * I feel like we need to define margins/padding for people with some explanatory text, at a least a little. Maybe something like "Margins (adjust spacing on the sides of the block)" and "Padding (adjust spacing around content within the block)." I don't know if the text goes under the terms themselves or if both definitions go at the bottom of the settings; that's a question for y'all.

<img width="572" alt="Screenshot 2020-02-24 at 15 46 35" src="https://user-images.githubusercontent.com/444434/75167163-de63ba80-571c-11ea-888f-9d3c3cc18acf.png">

@michelleweber Thanks for this. I've updated. I've added this as "Help" text below the input. Adding in the label itself didn't look good.

> * For the "vertical" point, I'd probably try to keep it really simple in the main settings, like "To add space above or below a block, insert a Spacer block.

@mapk @michelleweber I've ended up removing the vertical stuff. Not sure how important it is because it could affect things differently depending on which Theme is active.

> I was able to see immediate visual changes with every selection – really good stuff!

@mapk I'm glad you think this is ok. I think some experimental is required by the user when using this kind of setting. I've also added an animation to make it more perceivable (see gif in PR desc).

> Are those words clear enough for people? Do people know the difference? Would we want to include any graphic to help explain this like:

I don't know, but I think it's probably as clear as we're going to get right now. We've added explanatory text as well so hopefully folks with "get" it. We can always look at adding diagrams like you suggest in a follow-up PR.

> I didn't notice any change in the frontend yet. Is that correct? If I'm right that the frontend work just hasn't been done yet, it might make a difference to number 1 above.

@mapk I was testing using the Gutenbehr Starter Theme. I tested with Twenty Twenty and there is a CSS specificity issue. It looks like Twenty Twenty adds a `editor-styles-wrapper` class to the CSS that applies to backgrounds which adds a padding value (see below). This causes the "core" styles for padding that I've added to be overidden. Same for margin. Any ideas how we work around this kind of upstream issue?

![Screen Shot 2020-02-24 at 15 40 19](https://user-images.githubusercontent.com/444434/75167652-9729f980-571d-11ea-8647-ef23c9d7901b.png)

shaunandrews commented:
I wonder if we can simplify the spacing settings to reduce some of the copy. Maybe using "Inside" and "Outside" along with some visual explanation could help:

<img width="319" alt="image" src="https://user-images.githubusercontent.com/191598/75169494-a9447180-56f6-11ea-8d13-a9769a45b12b.png">

kjellr commented:
👋 Hey folks. Taking a look at this PR for the first time in a long time. A couple things: 

- I've tested this in both Twenty Nineteen and Twenty Twenty. It works nicely in the editor in both cases, but the front-end is mixed. The padding _sometimes_ works, but the margin is totally overwritten in both cases. Both would need patches.
- Because both themes rely on `auto` left/right margins for blocks inside of `entry-content`, I wouldn't expect that the left/right margin rules in here would ever actually work... so I'm not totally sure what to do about that. 😕
- While it's totally possible to tie these new spacing classnames into those themes, I'll note that it's going to require _immensely_ complicated CSS. Twenty Nineteen for instance relies on the padding value in its calculation for `max-width` [in a few different places](https://core.trac.wordpress.org/browser/trunk/src/wp-content/themes/twentynineteen/sass/blocks/_blocks.scss#L817), and will have to provide alternate calculations for each one of these size presets. 

That's all to say... themes can handle this, but it will require creating some relatively difficult patches. And I worry that it's just making an already-complicated approach to the Group block styles even more complicated for those themes. 

The best solution would be to figure out a way for core to handle all these wide/full/group/spacing stuff, but I think we're a long way off from that becoming a reality.

getdave commented:
@kjellr thanks for taking a look. I also noted some of your points during testing and didnt have a good solution

https://github.com/WordPress/gutenberg/pull/16730#issuecomment-590397567

Are we saying this isnt a goer? If so at least it has been explored and found wanting.

In my head Themes need to evolve to handle this kind of manipulation from the editor. I wonder if allowinh Themes to register custom padding and margin values and making these available as CSS vars would help. Then based on which padd margin value is set in Editor the Theme could adjust its stylesheet using the CSS vars.

lkraav commented:
> Are we saying this isnt a goer? If so at least it has been explored and found wanting.

I've been following along here as well, and pondering about whether to say anything in this direction or not. Sometimes it's useful to let things play out, since other innovations may blossom.

I think this type of block-level CSS micro-management is at risk of running into mountain of mismatch issues on the frontend across the theme universe. Containers, such as Group block, can have a wild array of different uses, each needing their own systematic spacing setup, unlike a blockquote element, for an example.

At minimum, these controls shouldn't be enabled by default - only themes declaring and having compatibility should enable them.

ZebulanStanphill commented:
I would say that it's bad practice for themes to center the content using auto margins, but it's not that easy to accomplish the same thing without either:

- giving up floats and using flex display mode
- using a container div for the content, using the margins on that, and then have to hackily implement wide/full widths using negative margins

So we definitely can't just ask themes to stop using the auto margin method.

But then again, is this even a problem? If there's nothing on the left or right, you don't really need to add margins there. The vertical margin will still be applied. Depending on how you look at it, this could even be considered a feature.

While I would have been more than happy to see this feature implemented a year ago, the Global Styles discussion (among other recent developments) has partially changed my mind about adding this kind of feature. I think it might be best to avoid adding any more new kinds of style controls until we know for sure how Global Styles is going to work. Otherwise, we may end up creating another legacy option/control/style/etc. that will hinder future progress with backwards compatibility concerns.

getdave commented:
Thanks all for the feedback and input.

Given that @mtias raised the original Issue, I'd really like to have his opinion on whether we continue to pursue this feature in this PR. Hopefully he can offer some direction based on his in-depth knowledge of where the Block Editor is heading.

> But then again, is the horizontal-margins-not-being-applied even a problem? If there's nothing on the left or right, you don't really need to add margins there. The vertical margin will still be applied. Depending on how you look at it, this could even be considered a feature.

Given how themes rely on auto margins, perhaps we could simply give up on margins for now and just pursue the padding element? 

> At minimum, these controls shouldn't be enabled by default - only themes declaring and having compatibility should enable them.

We could indeed make both dimension controls an opt-in feature thereby allowing Theme authors to choose to implement the necessary changes in their own time. For example, if you know your Themes relies on auto margins for centring then you might:

* not opt in to dimensions at all
* opt in to padding but not margin
* opt in to both but choose to override hoz margin with your own CSS styles 

This would avoid the "mountain of mismatch issues" [suggested above](https://github.com/WordPress/gutenberg/pull/16730#issuecomment-590498988).

That said, I am not wholly convinced that just because Theme authors have to make some changes we should limit the power exposed to users to control their website via the Editor. There's a delicate balance to be had, but I think with some careful planning we can mitigate any negative effects.

> While I would have been more than happy to see this feature implemented a year ago, the Global Styles discussion (among other recent developments) has partially changed my mind about adding this kind of feature. 

Thanks for raising this. Reading the comment, I was struggling to understand why Global Styles makes individual block controls redundant. Even if we have some kind of global styles that allows control over Block padding/margin then it's still likely users will want/need to control Block dimensions individually. If I've missed something please let me know, as I'm not super familiar with the progress of Global Styles at the moment.

lkraav commented:
> That said, I am not wholly convinced that just because Theme authors have to make some changes we should limit the power exposed to users to control their website via the Editor. There's a delicate balance to be had, but I think with some careful planning we can mitigate any negative effects.

The only negative effect really is the hodge-podge result across pages / blocks, since there is no "Global Styles" yet to keep things in sync, and micro-managing consistency across blocks is too much work for probably everyone (if it's even technically possible to manually curate a 100% coverage, due to lack of proper visibility tools, etc).

It's true, though, that sometimes adjusting a specific block like this achieves the user's goals much cheaper than finding/hiring someone to do a "proper" consistent job across the board. It's certainly the type of Power we'd like to provision, in general.

My "controlled enable/disable" thinking above is more suited for the scenario where higher fidelity solution resources are available, and editors should use them, instead of hacking individual blocks that nobody will have time to find and unhack later (temporary things are forever, right?).

kjellr commented:
Regarding Global Styles, @karmatosed may be able to fill in more details here. But in general, the benefit would be that there's a system set up to manage these sorts of settings and translate them so that themes can hook into them. 

As a simple example: If this spacing setting was recorded in a CSS variable via Global Styles, then themes could hook into that, and it would automatically adjust the spacing in real-time without requiring a bunch of extra classnames and complicated style rules. 

Looking further out, global styles (and full-site editing in general) also allows for the possibility that the basic stylesheet scaffolding of these blocks would be entirely handled by Gutenberg. In this case, the group block would work out of the box without the theme supplying front-end code. This is a little bit of a future wish on my part, but Global Styles and Full-site editing get us closer to that then we are today. If this were to happen, then we wouldn't have to worry about these sorts of theme issues at all moving forward.

mtias commented:
The lack of control over padding (in the group block in particular) is becoming a pretty severe limitation for building block patterns. We won't be able to solve all problems at once, but we should start by adding padding control to Group to allow these customizations. This is orthogonal to the global style system recognizing and setting defaults together with the theme which cascade through blocks.

@ItsJonQ was exploring having canvas controls for it as well: https://codesandbox.io/s/happy-platform-0leig

karmatosed commented:
Thanks for the ping @kjellr. I explored this a little for width and padding in global styles here: https://github.com/WordPress/gutenberg/issues/20771

getdave commented:
> The lack of control over padding (in the group block in particular) is becoming a pretty severe limitation for building block patterns
> We won't be able to solve all problems at once

Agreed 💯 

> we should start by adding padding control to Group to allow these customizations

Let's do this. @shaunandrews [presented a design](https://github.com/WordPress/gutenberg/pull/16730#issuecomment-590409843). @karmatosed would this work for you? How shall I proceed visually here? 

> @ItsJonQ was exploring having canvas controls for it as well: https://codesandbox.io/s/happy-platform-0leig

This looks great and would be vastly preferable. Are we agreed this would be something for a followup PR?

ItsJonQ commented:
> This looks great and would be vastly preferable. 
@getdave Thanks for the feedback! That was a really quick test. It sounds like there may be something to it. In that case, I can work on refining the prototype to make sure the interactions and intent feel right

ItsJonQ commented:
👋 Hai there!!

Apologies for the late reply! It's been a bit hectic.

Updates!

The `UnitControl` UI I prototyped is making it's way into core via this PR:
https://github.com/WordPress/gutenberg/pull/20888

The above PR is very close to landing.
Once it does, I'll be able to use it to create the dimension controls for blocks.

For context, it would be for the top / right / bottom / left based padding values. A demo of this can be found in my prototype here:

https://0leig.codesandbox.io/

(The padding units at the top)

getdave commented:
Ill put this back on my roadmap of issues to get done.

paaljoachim commented:
I just linked to another PR for adding padding to the Cover Block. Seen just above.
I would think that the method/component made in the PR for Cover Block padding should also transfer over the Group Block.

getdave commented:
> The above PR is very close to landing.
> Once it does, I'll be able to use it to create the dimension controls for blocks.

@ItsJonQ Sounds like we'll be able to merge that into this PR quite easily then? Core Navigation is taking all my time at the moment anyway, so I'll wait for you to finish up.

paaljoachim commented:
It looks like this PR has stalled. 
What is the forward path @getdave and @ItsJonQ ?

getdave commented:
@ItsJonQ I believe you've been iterating quite heavily on adding features to the Cover block. Are you able to briefly outline the steps that might need to be taken to add the dimension controls to the Group block as well? I'm happy to try and take a look at doing it, but it might save a tonne of time if you could roughly sketch out the roadmap as I'm slightly out of touch with latest UI control developments.

ItsJonQ commented:
> Are you able to briefly outline the steps that might need to be taken to add the dimension controls to the Group block as well? ...  roughly sketch out the roadmap .

@getdave Haiii!! I'll do my best to help 💪 

### Supports

With the `supports` feature, the Spacing (padding) controls can be added to any block (in theory) by specifying it in the `block.json`:

```
	"supports": {
		"anchor": true,
		"align": true,
		"html": false,
		"spacing": {
			"padding": true // 👈 Right here
		}
	}
```

The values are (currently) rendered in the block as inline styles.

### Controls

For the above system to work, the controls are "bound" to the support attributes through the `block-editor/hooks`:
https://github.com/WordPress/gutenberg/tree/master/packages/block-editor/src/hooks

Here are the padding controls:
https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/hooks/padding.js

These controls are abstracted away from a block's `edit.js` file (where controls have traditionally lived), in this "global" layer. This allows blocks to flag which controls they'd need, whether it's `padding` or `align`.

### Visualizers

The tricky part is, the Cover also adds some UI to help "visualize" the padding changes via these transparent blue rectangles. The implementation of that currently looks like this:

https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/cover/edit.js#L533

(I wish we didn't have to manually rig this part up. But that's how it currently has to be done)

The spacing controls have to be enabled via the (`experimental`) `theme.json` file in the theme, like so:

```js
// experimental-theme.json
{
  "global": {
    "settings": {
      "spacing": {
        "customPadding": true
      }
    }
  }
}

```

### Group block

It looks like the `spacing.padding` support has already been added to the "Group" block.

---

Hopefully the above explanation helps clarify some of the mechanics of how this is coordinated 🙏

getdave commented:
Now being handled in https://github.com/WordPress/gutenberg/pull/25988
